lib/connections/quic_misc.go

```diff
@@ -0,0 +1,141 @@
+// Copyright (C) 2019 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package connections
+
+import (
+	"bytes"
+	"net"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/lucas-clemente/quic-go"
+)
+
+var (
+	mut   sync.Mutex
+	conns []net.PacketConn
+)
+
+var (
+	stunFilterPriority = 10
+	quicFilterPriority = 100
+	quicConfig         = &quic.Config{
+		ConnectionIDLength: 4,
+		KeepAlive:          true,
+	}
+)
+
+// Track connections on which we are listening on, to increase the probability of accessing a connection that
+// has a NAT port mapping. This also makes our outgoing port stable and same as incoming port which should allow
+// better probability of punching through.
+func getListeningConn() net.PacketConn {
```

Comment:
I think this stuff needs handling for quic4/quic6, lest we have a listener `quic4://...` and try to use that socket to dial an IPv6 address, or vice versa, which will fail.